ports/stm32/uart.c

```diff
@@ -825,21 +836,6 @@ void uart_irq_handler(mp_uint_t uart_id) {
     }
     #endif
 
-    // Set user IRQ flags
-    self->mp_irq_flags = 0;
-    #if defined(STM32F4)
-    if (self->uartx->SR & USART_SR_IDLE) {
-        (void)self->uartx->SR;
-        (void)self->uartx->DR;
```

Comment:
I believe this is a bug in the existing implementation, which still occurs in my changes. In the case where the buffer is already full, the data in `DR` will not have been read above and will still be sitting there. That data will now be cleared/lost here if the IDLE flag is also set.

I guess the reality in any uart system is if the buffer isn't read fast enough data is being lost, so this is no different. It might be nice to detect this and set a flag the user can see though?

Other than this, I do believe this PR is good for final review, I've been using it unchanged for a long time now on our project.

Reply:
You're right, it was broken before this PR!  Eg if `rxbuf=0` then there is no buffer and it is supposed to allow 1 character reads, but with irqs enabled all chars disappear.